### DISCUSSION TITLE
Better understanding of system design choices

### DISCUSSION THREAD

**Discussion by sumitbindra** [langroid/langroid#510] [Q&A] [answered]
Hello,

I came across langroid yesterday and it caught my attention. Read through the examples and documentations but still have a some questions. Hopefully some others might also find them useful.

A simple example that I will use here is creating an tech development agency with CEO, PM and developer.

1. Human intervention: Why does the human have to respond after each step (in most examples it is by pressing enter)? Say I am the (human) client and I launched this agency to write software for me. Once I give the CEO the task, why do I (the human) have to interact by pressing enter to keep the task going? 
POTENTIAL SOLUTION: I think this is the "interactive" flag in the task. Had to go find this one. 
FOLLOW UP: Was the default choice mainly to prevent infinite loops during testing? 

2. Task Delegation: The documentation mentions this default order of task delegation 

[self.agent_response, self.llm_response, self.user_response] 

I think this means that first the task is sent to the default agent which in itself has same three responders. When you add sub-tasks the order becomes:

[self.agent_response, self.llm_response, self.user_response, pm_task.run()]

In the example below, I the human client gives the CEO instructions to create a program. What I think is going to happen is the CEO will first look at the CEO_agent to see if the agent can do it. Say we instruct it to not write any programs and it doesn't. Next is the LLM of the task which in most cases will be capable of writing the program. Does it inherit the instructions of the agent and also decides to not write the program? 

The goal of an agent like the CEO is to communicate with client and PM (via ceo_agent.enable_message(RecipientTool)) to gather requirements and assign a deadline, does it make sense to put the pm_task at the end. I noticed that if I didn't explicitly state that "DONT WRITE ANY PROGRAM" , the LLM in the CEO agent started writing the code without delegating it to the PM potentially due to this task delegation order.

3. Conflict with enabling recipient tool (for multi user collar) and llm_delegate = True for using the llm to decide when the task done is a bit confusing. 

4. Is there any difference between: 

ceo_task.add_sub_task(pm_task)
pm_task.add_sub_task(dev_task)

ceo_task.run()

vs 

pm_task.add_sub_task(dev_task)
ceo_task.add_sub_task(pm_task)

ceo_task.run()

It felt like the results weren't consistent between the two. 

thank you for creating this and I hope some of these questions help bring along other novice users. 

```
import langroid as lr
import langroid.language_models as lm
from langroid.agent.tools.recipient_tool import RecipientTool

# set up LLM
llm_cfg = lm.OpenAIGPTConfig( # or OpenAIAssistant to use Assistant API 
  # any model served via an OpenAI-compatible API
  chat_model=lm.OpenAIChatModel.GPT4_TURBO, # or, e.g., "ollama/mistral"
)

# use LLM in an Agent
agent_cfg = lr.ChatAgentConfig(llm=llm_cfg)

# Multi-Agent chat loop: Teacher Agent asks questions to Student Agent

ceo_agent = lr.ChatAgent(agent_cfg)
ceo_agent.enable_message(RecipientTool)  # because it is talking to more than one agent? Human client and PM
ceo_task = lr.Task(
  ceo_agent, name="CEO",
  llm_delegate=True,    # because I want to LLM to decide when the task is done instead of some special meesage
  system_message="""
    You are the CEO of a software developement company called AdvancedTech focused on understanding client needs, 
    knowing your team strengths and delivering a product that solves the clients problems completely and thoroughly. 
    
    You communicate ONLY with the client and your projet manager.

    YOU MUST ALWAYS START BY ASKING THE CLIENT WHAT THEY NEED.

    DONT WRITE ANY PROGRAM. 
    
    INSTEAD SEND THE TASK TO THE PROJECT MANAGER AS IS WITH A REASONABLE DEADLINE.
    """
)

pm_agent = lr.ChatAgent(agent_cfg)
pm_agent.enable_message(RecipientTool)  # because it is talking to more than one agent? Talking to CEO and developer
pm_task = lr.Task(
  pm_agent, name="Project Manager",
  llm_delegate=True,    # because I want to LLM to decide when the task is done instead of some special meesage
  interactive=False,
  system_message="""
    You are the lead project manager at a software developement company called AdvancedTech focused on receiving the 
    requirments of a project from the CEO and working with you tech developer to deliver the client's program.
    You communicate both with the CEO and the developer. You do not directly talk to the client.
    """
)

dev_agent = lr.ChatAgent(agent_cfg)
#dev_agent.enable_message(RecipientTool)  # because it is NOT talking to more than one agent? Talking to ONLY PM
dev_task = lr.Task(
  dev_agent, name="Developer",
  interactive=False,
  system_message="You are an expert python developer.",
  single_round=True
)


ceo_task.add_sub_task(pm_task)
pm_task.add_sub_task(dev_task)

ceo_task.run()
```

**Comment by pchalasani** [comment]
@sumitbindra thanks for trying out Langroid. 

> 1. Was the default choice mainly to prevent infinite loops during testing?

Yes exactly

For the ceo example, there are a few things to note:

- agent name cannot have spaces (this is documented in the api docs)
- in addition to enabling `RecipientTool` for an agent, you have to instruct the agent about who it can address. See several examples in the repo that do this. 
- `Client` is not a defined agent; instead it should address `user`
- `agent_response` is typically just a tool handler, so in this case it just handles routing when `RecipientTool` is used.

**Comment by pchalasani** [comment]
> first look at the CEO_agent to see if the agent can do it. Say we instruct it to not write any programs and it doesn't. Next is the LLM of the task which in most cases will be capable of writing the program. Does it inherit the instructions of the agent and also decides to not write the program?

As mentioned above, `agent_response` just handles tools, so the LLM is the one that follows instructions set in the `system_message`. 

> does it make sense to put the pm_task at the end. I noticed that if I didn't explicitly state that "DONT WRITE ANY PROGRAM" , the LLM in the CEO agent started writing the code without delegating it to the PM potentially due to this task delegation order.

Yes the way the task loop works is that in each `step()` it sequentially tries to find a valid response from one of the responders in the responder-list, and as soon as there's a valid response, the `step()` is finished, and in the next step, the search for valid response starts _at the beginning of the list_, subject to rules like an entity/responder cannot respond to its own message, etc. In this way after the `llm_response` of CEO says something to the PM, the `PM.run()` will be the only one eligible to handle the message.

You can look at the logs under `logs/` to get an idea of what is going on. Logs are explained in the `Task delegation` section of the docs.

> * Conflict with enabling recipient tool (for multi user collar) and llm_delegate = True for using the llm to decide when the task done is a bit confusing.
> *

These are distinct constructs. Can you elaborate on the confusion?


> It felt like the results weren't consistent between the two.

Should be essentially same. The variation you are seeing could be due to sampling difference on different runs (even with Temp=0). If you see substantial variation let me know.

**Comment by pchalasani** [comment, marked as answer]
Here is a version of the code that should work as expected. I'm pretty sure there are examples in the `examples` directory where the sub-task structure is of this exact form, i.e. `Main->A->B`, you'll probably find them.

```
import langroid as lr
import langroid.language_models as lm
from langroid.agent.tools.recipient_tool import RecipientTool

# set up LLM
llm_cfg = lm.OpenAIGPTConfig(  # or OpenAIAssistant to use Assistant API
    # any model served via an OpenAI-compatible API
    chat_model=lm.OpenAIChatModel.GPT4_TURBO,  # or, e.g., "ollama/mistral"
)

# use LLM in an Agent
agent_cfg = lr.ChatAgentConfig(llm=llm_cfg)

# Multi-Agent chat loop: Teacher Agent asks questions to Student Agent

ceo_agent = lr.ChatAgent(agent_cfg)
ceo_agent.enable_message(
    RecipientTool
)  # because it is talking to more than one agent? Human client and PM
ceo_task = lr.Task(
    ceo_agent,
    name="CEO",
    interactive=False,
    llm_delegate=True,  # because I want to LLM to decide when the task is done instead of some special meesage
    system_message="""
    You are the CEO of a software development company called AdvancedTech focused on understanding client needs, 
    knowing your team strengths and delivering a product that solves the clients problems completely and thoroughly. 
    
    You communicate ONLY with the User and your ProjectManager.
    Use the `recipient_tool` to address one of these.

    YOU MUST ALWAYS START BY ASKING THE CLIENT WHAT THEY NEED.

    DONT WRITE ANY CODE.
    
    INSTEAD SEND THE TASK TO THE PROJECT MANAGER AS IS WITH A REASONABLE DEADLINE.
    
    Once you receive the code from the Project Manager, 
    simply say DONE.
    """,
)

pm_agent = lr.ChatAgent(agent_cfg)
pm_agent.enable_message(
    RecipientTool
)  # because it is talking to more than one agent? Talking to CEO and developer
pm_task = lr.Task(
    pm_agent,
    name="ProjectManager",
    llm_delegate=True,  # because I want to LLM to decide when the task is done instead of some special meesage
    interactive=False,
    system_message="""
    You are the lead project manager at a software development company 
    called AdvancedTech focused on receiving the 
    requirments of a project from the CEO and working with you tech developer to 
    deliver the client's program.
    You communicate only with either with the "CEO" or "Developer".
    Use the `recipient_tool` to address one of these. 
    
    As soon as you get code from the developer, address the CEO 
    and show them the code. 
    """,
)

dev_agent = lr.ChatAgent(agent_cfg)
# dev_agent.enable_message(RecipientTool)  # because it is NOT talking to more than one agent? Talking to ONLY PM
dev_task = lr.Task(
    dev_agent,
    name="Developer",
    interactive=False,
    system_message="""
    You are an expert python developer. Simply return code for the given task.
    """,
    single_round=True,
)


ceo_task.add_sub_task(pm_task)
pm_task.add_sub_task(dev_task)

ceo_task.run()

```

**Comment by sumitbindra** [comment]
Tested it out and worked as expected. Thank you. 

Will be playing more with it in coming days.

**Reply by pchalasani** [reply]
Also look at this example:
https://github.com/langroid/langroid/blob/main/examples/basic/chat-tree.py

And described here https://langroid.github.io/langroid/examples/agent-tree/
